src/webgpu/api/validation/validation_test.ts

```diff
@@ -1,17 +1,103 @@
+import { attemptGarbageCollection } from '../../../common/util/collect_garbage.js';
 import { assert, unreachable } from '../../../common/util/util.js';
-import { BindableResource, kMaxQueryCount } from '../../capability_info.js';
+import { ValidBindableResource, BindableResource, kMaxQueryCount } from '../../capability_info.js';
 import { GPUTest } from '../../gpu_test.js';
+import {
+  DevicePool,
+  DeviceProvider,
+  TestOOMedShouldAttemptGC,
+  UncanonicalizedDeviceDescriptor,
+} from '../../util/device_pool.js';
 
 import { CommandBufferMaker, EncoderType } from './util/command_buffer_maker.js';
 
 const kResourceStateValues = ['valid', 'invalid', 'destroyed'] as const;
 export type ResourceState = typeof kResourceStateValues[number];
 export const kResourceStates: readonly ResourceState[] = kResourceStateValues;
 
+const mismatchedDevicePool = new DevicePool();
```

Comment:
I see the device pool caches the device providers by the descriptor,  but we need to provide two devices with same descriptor at once. Do you mean a device provider could provide multiple devices?

Reply:
Eventually the plan for the DevicePool is that it'll be able to provide multiple devices concurrently (hence the name "pool"), so that multiple test cases can run concurrently within the same page. That will also allow a single case (like these) to get multiple devices from the pool.